### PR TITLE
[OHIF-177] - First pass at SR STOW with Arrow Annotate, Ellipse, Length and Bidirectional.

### PR DESCRIPTION
**Pull request by JamesAPetts**
When you click the Create Report button, we should serialize all tracked measurements to DICOM SR and STOW to PACS.

I also have local download as that was the first test, currently bound to export, but with no fancy UI yet.

No UI for successful post yet. After posting refresh the page after a few seconds and you can view the SR with the SR viewport.

### CODE REVIEW

File: extensions/default/src/Panels/WrappedPanelStudyBrowser.jsx [hunk 1]
```diff
@@ -20,7 +20,7 @@ function WrappedPanelStudyBrowser({
 }) {
   // TODO: This should be made available a different way; route should have
   // already determined our datasource
-  const dataSource = extensionManager.getDataSources('dicomweb')[0];
+  const dataSource = extensionManager.getDataSources()[0];
```
Comment (dannyrb) [inline extensions/default/src/Panels/WrappedPanelStudyBrowser.jsx:23]: Didn't you add some magic way to get the "default" data source? Or doesn't the mode route publish/identify the selected data source in some way?

Comment (JamesAPetts) [reply extensions/default/src/Panels/WrappedPanelStudyBrowser.jsx:23]: If you don't give it a tag you get the default. We could add an explicit proxy for the same function also, just saw this explictly assumption and nixxed it as I walked by.

File: platform/core/src/DataSources/IWebApiDataSource.js [hunk 2]
```diff
@@ -44,13 +45,20 @@ function create({
...
+      throw new Error(
+        'store.dicom(naturalizedDicom, StudyInstanceUID) not implemented for dataSource.'
+      );
+    },
+  };
+
   return {
     query: query || defaultQuery,
     retrieve: retrieve || defaultRetrieve,
+    store: store || defaultStore,
```
Comment (dannyrb) [inline platform/core/src/DataSources/IWebApiDataSource.js:59]: We really need to discuss the shape of this API pre big release.
This follows the current pattern. Data sources are weird.

File: extensions/default/src/Panels/WrappedPanelStudyBrowser.jsx (hunk 1 above)
Comment (dannyrb) [reply extensions/default/src/Panels/WrappedPanelStudyBrowser.jsx:23]: So it returns a list of "active data sources", with the first being the default? Providing a name filters, but still gives a list?

A proxy would be neat. If there's enough here for a follow-up issue, great. If not, just resolve this comment thread.

File: extensions/default/src/DicomWebDataSource/index.js [hunk 3]
```diff
@@ -128,6 +133,32 @@ function createDicomWebApi(dicomWebConfig) {
...
+
+        const part10Buffer = dicomDict.write();
+
+        const options = {
+          datasets: [part10Buffer],
+        };
+
+        await wadoDicomWebClient.storeInstances(options);
+      },
+    },
```
Comment (dannyrb) [inline extensions/default/src/DicomWebDataSource/index.js:161]: You make it look so easy 😎

File: extensions/measurement-tracking/src/panels/PanelMeasurementTableTracking/index.js [hunk 4]
```diff
@@ -102,6 +105,37 @@ function PanelMeasurementTableTracking({ servicesManager, commandsManager }) {
...
+  const onCreateReportClick = () => {
+    const measurements = MeasurementService.getMeasurements();
+    const trackedMeasurements = measurements.filter(
+      m =>
+        trackedStudy === m.referenceStudyUID &&
+        trackedSeries.includes(m.referenceSeriesUID)
+    );
+
+    console.log(servicesManager);
+    console.log(extensionManager);
```
Comment (dannyrb) [inline extensions/measurement-tracking/src/panels/PanelMeasurementTableTracking/index.js:129]:
```suggestion
```

File: extensions/measurement-tracking/src/panels/PanelMeasurementTableTracking/index.js [hunk 5]
```diff
@@ -102,6 +105,36 @@ function PanelMeasurementTableTracking({ servicesManager, commandsManager }) {
...
+
+  const onCreateReportClick = () => {
+    const measurements = MeasurementService.getMeasurements();
+    const trackedMeasurements = measurements.filter(
+      m =>
+        trackedStudy === m.referenceStudyUID &&
+        trackedSeries.includes(m.referenceSeriesUID)
+    );
+
+    console.log(servicesManager);
```
Comment (dannyrb) [inline extensions/measurement-tracking/src/panels/PanelMeasurementTableTracking/index.js:128]:
```suggestion
```

File: extensions/measurement-tracking/src/panels/PanelMeasurementTableTracking/index.js [hunk 6]
```diff
@@ -102,6 +105,35 @@ function PanelMeasurementTableTracking({ servicesManager, commandsManager }) {
 
   const activeMeasurementItem = 0;
 
+  const onExportClick = () => {
```
Comment (dannyrb) [inline extensions/measurement-tracking/src/panels/PanelMeasurementTableTracking/index.js:108]: `onExportClick` is the name of the prop. For it to read smoother you can name this `localDownloadStructuredReport`

So the prop binding reads:

`onExportClick={localDownloadStructredReport}`

Comment (dannyrb) [reply extensions/measurement-tracking/src/panels/PanelMeasurementTableTracking/index.js:108]: If this doesn't depend on reactive values, you don't need to define this function in the component. Re-renders will recreate it. You can shift it outside of the component in the same file, or a separate file.

File: platform/core/src/DICOMSR/dataExchange.js [hunk 7]
```diff
@@ -78,4 +82,112 @@ const storeMeasurements = async (measurementData, filter, server) => {
   }
 };
 
-export { retrieveMeasurements, storeMeasurements };
+const downloadReport = measurementData => {
```
Comment (dannyrb) [inline platform/core/src/DICOMSR/dataExchange.js:85]: JS Doc or description of what `measurementData` is?

File: platform/core/src/DICOMSR/dataExchange.js [hunk 8]
```diff
@@ -78,4 +82,112 @@ const storeMeasurements = async (measurementData, filter, server) => {
...
+
+  const report = MeasurementReport.generateReport(
+    filteredToolState,
+    cornerstone.metaData
+  );
+
+  return report.dataset;
+};
+
+const storeMeasurements = async (measurementData, dataSource) => {
```
Comment (dannyrb) [inline platform/core/src/DICOMSR/dataExchange.js:106]: Same here. If this takes a `dataSource`, can we call out the required props in JSDoc?

`@param {func} dataSource.store.dicom - does stuff`

File: platform/core/src/DICOMSR/dataExchange.js [hunk 9]
```diff
@@ -78,4 +82,112 @@ const storeMeasurements = async (measurementData, filter, server) => {
...
+    }
+
+    return {
+      message: 'Measurements saved successfully',
+    };
+  } catch (error) {
+    log.error(
+      `[DICOMSR] Error while saving the measurements: ${error.message}`
+    );
+    throw new Error('Error while saving the measurements.');
```
Comment (dannyrb) [inline platform/core/src/DICOMSR/dataExchange.js:133]: If you haven't already, separate issue calling out UI notification for success/fail.
Outside of sprint, blocked by design/user/func requirements

File: extensions/default/src/Panels/WrappedPanelStudyBrowser.jsx (hunk 1 above)
Comment (JamesAPetts) [reply extensions/default/src/Panels/WrappedPanelStudyBrowser.jsx:23]: It returns the active data source, the datasource might have multiple servers. In our example it returns the list of dicomweb datasources. We have the [0] everywhere throughout the code so we could build with multiple in mind ^^.

Comment (dannyrb) [reply extensions/default/src/Panels/WrappedPanelStudyBrowser.jsx:23]: Ah! Thanks for the refresher.

File: platform/core/src/DataSources/IWebApiDataSource.js (hunk 2 above)
Comment (JamesAPetts) [reply platform/core/src/DataSources/IWebApiDataSource.js:59]: Agreed. I at least chose not to call it stow as that is DICOMWeb specific.

File: platform/core/src/DICOMSR/dataExchange.js (hunk 9 above)
Comment (dannyrb) [reply platform/core/src/DICOMSR/dataExchange.js:133]: https://precisionmetrics.atlassian.net/secure/RapidBoard.jspa?rapidView=1&projectKey=OHIF&modal=detail&selectedIssue=OHIF-177

**Comment by dannyrb** [comment]
Merging for now. Follow-up on anything you feel is relevant.

**Pull request merged**